### PR TITLE
Upgrade mkldnn-bridge for dnnlowp support

### PR DESCRIPTION
**Pull request by gujinghui**
The mkldnn-bridge is upgraded in this PR to support DNNLOWP operators.
Meanwhile, APIs have been updated in caffe2 to use latest version.

### CODE REVIEW

**Comment by gujinghui** [comment]
@jgong5 @Jianhui-Li

**Comment by gujinghui** [comment]
@yinghai 
pls help review.

**Comment by yinghai** [comment]
Can we have an RFC issue posted on what the motivation of this series of PRs is?

**Comment by Jianhui-Li** [comment]
Hi @yinghai I have posted an issue:  https://github.com/pytorch/pytorch/issues/16361  
Thanks!

File: caffe2/ideep/ideep_utils.h [hunk 1]
```diff
@@ -9,12 +9,23 @@
 
 namespace caffe2 {
 
+enum FusionType {
+  FUSION_UNKNOWN = 0,
+  FUSION_CONV_RELU = 1,
+  FUSION_CONV_SUM = 2,
+  FUSION_CONV_SUM_RELU = 3,
+  FUSION_MAX = FUSION_CONV_SUM_RELU + 1,
+};
```
Comment (gujinghui) [inline caffe2/ideep/ideep_utils.h]: Move enum from ConvFusion to here as public type.
Then, we can use it in optimize_ideep.cc

File: caffe2/ideep/ideep_utils.h [hunk 2]
```diff
@@ -9,12 +9,23 @@
...
 #define USE_IDEEP_DEF_ALIASES()                                                \
+  using ikey = ideep::key_t;                                                   \
   using itensor = ideep::tensor;                                               \
   using iformat = ideep::format;                                               \
+  using iscale = ideep::scale_t;                                               \
   using ialgo = ideep::algorithm;                                              \
   using iprop = ideep::prop_kind;                                              \
   using ipadding = ideep::padding_kind;                                        \
+  using idtype = ideep::tensor::data_type;                                     \
   using iattr = ideep::descriptor_group::attr_t;                               \
```
Comment (gujinghui) [inline caffe2/ideep/ideep_utils.h]: Expose new type definition to caffe2 for DNNLOWP support.

File: caffe2/ideep/operators/conv_fusion_op.cc [hunk 3]
```diff
@@ -7,40 +7,37 @@ class IDEEPConvFusionOp final : public IDEEPConvPoolOpBase {
   USE_IDEEP_DEF_ALIASES();
   USE_IDEEP_CONV_POOL_BASE_FUNCTIONS();
 
-  enum FusionType {
-    FUSION_UNKNOWN = 0,
-    FUSION_CONV_RELU = 1,
-    FUSION_CONV_SUM = 2,
-    FUSION_CONV_SUM_RELU = 3,
-    FUSION_MAX = FUSION_CONV_SUM_RELU + 1,
-  };
```
Comment (gujinghui) [inline caffe2/ideep/operators/conv_fusion_op.cc]: Moved to ideep_utils.h

File: caffe2/ideep/operators/conv_fusion_op.cc [hunk 4]
```diff
@@ -82,15 +68,22 @@ class IDEEPConvFusionOp final : public IDEEPConvPoolOpBase {
...
-              filter.get_dims());
+              filter.get_dims(),
+              filter.get_data_type(),
+              stride_,
+              pad_tl(),
+              pad_br(),
+              dilation_,
+              group_,
+              algo_,
+              pk_);
```
Comment (gujinghui) [inline caffe2/ideep/operators/conv_fusion_op.cc]: To query correct descriptor, more parameters are needed here.

File: caffe2/ideep/operators/conv_fusion_op.cc [hunk 5]
```diff
@@ -82,15 +68,22 @@ class IDEEPConvFusionOp final : public IDEEPConvPoolOpBase {
...
       if (filter_.get_descriptor() != expected_descriptor) {
-        filter_.init<ideep::utils::allocator, ideep::convolution_forward>(
-            expected_descriptor);
-        ideep::reorder::compute(filter, filter_);
+        filter_.init(expected_descriptor);
+        filter_.feed_from(filter);
```
Comment (gujinghui) [inline caffe2/ideep/operators/conv_fusion_op.cc]: reorder is obseleted. use feed_from API in caffe2.

File: caffe2/ideep/operators/conv_fusion_op.cc [hunk 6]
```diff
@@ -102,7 +95,9 @@ class IDEEPConvFusionOp final : public IDEEPConvPoolOpBase {
           pad_tl(),
           pad_br(),
           group_,
-          attr());
+          attr_,
+          algo_,
+          pk_);
```
Comment (gujinghui) [inline caffe2/ideep/operators/conv_fusion_op.cc]: Conv API is extended. More parameters are needed.

File: caffe2/ideep/operators/conv_fusion_op.cc [hunk 7]
```diff
@@ -127,6 +124,10 @@ class IDEEPConvFusionOp final : public IDEEPConvPoolOpBase {
   }
 
  private:
+  iprop pk_;
+  ialgo algo_;
+  iattr attr_;
+  int last_input_;
```
Comment (gujinghui) [inline caffe2/ideep/operators/conv_fusion_op.cc]: Cached these values here. Then, no need to initialize them every time.

File: caffe2/ideep/operators/conv_pool_base_op.h [hunk 8]
```diff
@@ -43,7 +40,7 @@ class IDEEPConvPoolOpBase : public ConvPoolOpBase<IDEEPContext> {
     InferOutputSize(
         input_Tdims,
         output_channel,
-        order_,
+        StorageOrder::NCHW, //order_,
```
Comment (gujinghui) [inline caffe2/ideep/operators/conv_pool_base_op.h]: In MKL-DNN, the shape is always unchanged, no matter what kind of memory layout the tensor is taken.
In other words, the order of shape is always same as NCHW.
Therefore, the order_ of cafffe2 operator should not change the result of InferOutputSize.

File: caffe2/ideep/operators/operator_fallback_ideep.h [hunk 9]
```diff
@@ -95,7 +95,7 @@ class C10_EXPORT IDEEPFallbackOp final : public IDEEPOperator {
           dtensor->ShareExternalPointer(
               static_cast<float*>(input.get_data_handle()));
         } else {
-          input.reorder_to(dtensor->template mutable_data<float>());
+          input.to_public(dtensor->template mutable_data<float>());
```
Comment (gujinghui) [inline caffe2/ideep/operators/operator_fallback_ideep.h]: the reorder_to is obselted. use to_public here.

File: caffe2/ideep/ideep_utils.h [hunk 10]
```diff
@@ -24,6 +24,7 @@ enum FusionType {
   using iscale = ideep::scale_t;                                               \
   using ialgo = ideep::algorithm;                                              \
   using iprop = ideep::prop_kind;                                              \
+  using ilowp_kind = ideep::lowp_kind;                                         \
```
Comment (gujinghui) [inline caffe2/ideep/ideep_utils.h]: expose the new type definition for DNNLOWP support

File: caffe2/ideep/utils/ideep_operator.h [hunk 11]
```diff
@@ -109,4 +109,14 @@ class IDEEPOperator : public OperatorBase {
...
 
+// Convert zero_point scales to min_max scales
+inline ideep::scale_t ConvertScales(
+    const std::vector<float> scales_z) {
+  ideep::scale_t scales (scales_z);
+  for (auto it = scales.begin(); it != scales.end(); it++) {
+    *it = 1.0f / *it;
+  }
+  return scales;
+}
```
Comment (gujinghui) [inline caffe2/ideep/utils/ideep_operator.h]: Convert the scale from FBGEMM scale to MKL-DNN scale.

File: caffe2/ideep/operators/conv_fusion_op.cc [hunk 12]
```diff
@@ -127,6 +124,10 @@ class IDEEPConvFusionOp final : public IDEEPConvPoolOpBase {
   }
 
  private:
+  iprop pk_;
```
Comment (yinghai) [inline caffe2/ideep/operators/conv_fusion_op.cc]: Better name to show what this is? And please add comments to document these variables.

File: caffe2/ideep/utils/ideep_operator.h [hunk 13]
```diff
@@ -111,4 +109,14 @@ class IDEEPOperator : public OperatorBase {
       : IDEEPOperator(operator_def, ws) {}                                     \
   virtual ~name() {}
 
+// Convert zero_point scales to min_max scales
+inline ideep::scale_t ConvertScales(
+    const std::vector<float> scales_z) {
```
Comment (yinghai) [inline caffe2/ideep/utils/ideep_operator.h]: Const reference?

File: caffe2/ideep/utils/ideep_operator.h [hunk 14]
```diff
@@ -111,4 +109,14 @@ class IDEEPOperator : public OperatorBase {
...
+  ideep::scale_t scales (scales_z);
+  for (auto it = scales.begin(); it != scales.end(); it++) {
```
Comment (yinghai) [inline caffe2/ideep/utils/ideep_operator.h]: for (const auto i: scales)

File: caffe2/ideep/utils/ideep_operator.h [hunk 15]
```diff
@@ -111,4 +109,14 @@ class IDEEPOperator : public OperatorBase {
       : IDEEPOperator(operator_def, ws) {}                                     \
   virtual ~name() {}
 
+// Convert zero_point scales to min_max scales
```
Comment (yinghai) [inline caffe2/ideep/utils/ideep_operator.h]: Document why this conversion is needed?

**Review by yinghai — COMMENTED** [review]
Since we provide option to build int8 support in MKL, shall we also build these int8 ops optionally?

File: caffe2/ideep/operators/conv_fusion_op.cc (hunk 12 above)
Comment (gujinghui) [reply caffe2/ideep/operators/conv_fusion_op.cc]: Sure. Will describe the usage of all types introduced by ideep in ideep_utils.h.

File: caffe2/ideep/utils/ideep_operator.h (hunk 13 above)
Comment (gujinghui) [reply caffe2/ideep/utils/ideep_operator.h]: Yes. Unchangeable. The return result gives conversion output.

Comment (gujinghui) [reply caffe2/ideep/utils/ideep_operator.h]: Sure. Will correct it.

File: caffe2/ideep/utils/ideep_operator.h (hunk 14 above)
Comment (gujinghui) [reply caffe2/ideep/utils/ideep_operator.h]: OK. will correct it.
Not sure if foreach is stable before.

File: caffe2/ideep/utils/ideep_operator.h (hunk 15 above)
Comment (gujinghui) [reply caffe2/ideep/utils/ideep_operator.h]: Sure. Will add more comments.

**Comment by gujinghui** [comment]
> Since we provide option to build int8 support in MKL, shall we also build these int8 ops optionally?

I'm not sure whether it's needed or not.
For FBGEMM, it depends on the compiler and avx512 support.
But for MKL-DNN, the reference path is always workable.
So, the option is not necessary.

File: caffe2/ideep/operators/conv_op.cc [hunk 16]
```diff
@@ -8,12 +8,17 @@ class IDEEPConvOp final : public IDEEPConvPoolOpBase {
...
+      : IDEEPConvPoolOpBase(operator_def, ws) {
+    OPERATOR_NEEDS_FEATURE(order_ == StorageOrder::NCHW,
+        "Unsupported storage order.");
     OPERATOR_NEEDS_FEATURE(
         pad_l() == pad_r() && pad_t() == pad_b(),
         "Uneven padding not supported.");
+
+    attr_ = iattr();
+    algo_ = ialgo::convolution_direct;
+    training_mode_ = OperatorBase::GetSingleArgument<int>("training_mode", 0);
```
Comment (yinghai) [inline caffe2/ideep/operators/conv_op.cc]: Why don't we put this to the ctor initialization list as before?

**Comment by yinghai** [comment]
@gujinghui  which commit number of ideep are we updating to?

File: caffe2/ideep/operators/conv_fusion_op.cc [hunk 17]
```diff
@@ -7,40 +7,37 @@ class IDEEPConvFusionOp final : public IDEEPConvPoolOpBase {
...
-            OperatorBase::GetSingleArgument<int>("training_mode", 0)) {
+      : IDEEPConvPoolOpBase(operator_def, ws) {
+    OPERATOR_NEEDS_FEATURE(order_ == StorageOrder::NCHW,
+        "Unsupported storage order.");
     OPERATOR_NEEDS_FEATURE(
         pad_l() == pad_r() && pad_t() == pad_b(),
         "Uneven padding not supported.");
     OPERATOR_NEEDS_FEATURE(group_ == 1, "Group not supported.");
+
+    fusion_type_ = static_cast<FusionType>(
```
Comment (yinghai) [inline caffe2/ideep/operators/conv_fusion_op.cc]: same here.

**Review by yinghai — CHANGES_REQUESTED** [review]
Please rebase.

File: caffe2/ideep/operators/conv_op.cc (hunk 16 above)
Comment (gujinghui) [reply caffe2/ideep/operators/conv_op.cc]: Trying to merge Conv and ConvFusion operators in order to removing redundant code, in #16309.
To remove compiling warning, move out of ctor init list.

File: caffe2/ideep/operators/conv_fusion_op.cc (hunk 17 above)
Comment (gujinghui) [reply caffe2/ideep/operators/conv_fusion_op.cc]: Same as above.

**Comment by gujinghui** [comment]
> @gujinghui which commit number of ideep are we updating to?

@yinghai 
Will rebase and upgrade ideep.
Pls hold on.

**Comment by gujinghui** [comment]
@yinghai 
pls help review. depends on #9767.
pls merge #9767 firstly. I'll rebase this PR after #9767.

Thanks.

**Comment by gujinghui** [comment]
@yinghai
rebased. pls help review.

**Comment by gujinghui** [comment]
@yinghai
These two failures should be not related to this PR.

> 14:22:28 ======================================================================
14:22:28 ERROR: test_serialization_sharing (__main__.TestScript)
14:22:28 ----------------------------------------------------------------------
14:22:28 Traceback (most recent call last):
14:22:28   File "test_jit.py", line 102, in TemporaryFileName
14:22:28     yield f.name
14:22:28   File "test_jit.py", line 10613, in test_serialization_sharing
14:22:28     pickled_data = archive.read(os.path.join(archive_name, 'attributes.pkl'))
14:22:28   File "C:\Jenkins\Miniconda3\lib\zipfile.py", line 1337, in read
14:22:28     with self.open(name, "r", pwd) as fp:
14:22:28   File "C:\Jenkins\Miniconda3\lib\zipfile.py", line 1375, in open
14:22:28     zinfo = self.getinfo(name)
14:22:28   File "C:\Jenkins\Miniconda3\lib\zipfile.py", line 1304, in getinfo
14:22:28     'There is no item named %r in the archive' % name)
14:22:28 KeyError: "There is no item named 'tmphjchn_oh\\\\attributes.pkl' in the archive"


> DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-trusty-py3.6-gcc7:291
docker: Error response from daemon: failed to create endpoint condescending_kowalevski on network bridge: failed to add the host (veth9df0b97) <=> sandbox (vethdbcff1c) pair interfaces: operation not supported.
Warning: Permanently added the RSA host key for IP address '192.30.253.112' to the list of known hosts.
/bin/bash: line 11:  1164 Exit 1                  unbuffer bash ./command.sh
      1165 Segmentation fault      | ts
Exited with code 139

**Comment by yinghai** [comment]
@BIT-silence Could you take a look?

File: caffe2/ideep/ideep_utils.h [hunk 18]
```diff
@@ -15,13 +15,38 @@ enum ConvAlgorithm {
   CONV_ALGORITHM_MAX = CONV_ALGORITHM_WINOGRAD + 1
 };
 
+enum FusionType {
+  FUSION_UNKNOWN = 0,
+  FUSION_CONV_RELU = 1,
+  FUSION_CONV_SUM = 2,
+  FUSION_CONV_SUM_RELU = 3,
+  FUSION_MAX = FUSION_CONV_SUM_RELU + 1
```
Comment (BIT-silence) [inline caffe2/ideep/ideep_utils.h]: Why use FUSION_CONV_SUM_RELU + 1 here?

File: caffe2/ideep/operators/conv_op.cc [hunk 19]
```diff
@@ -2,116 +2,252 @@
...
+    CAFFE_ENFORCE(OperatorBase::HasArgument("fusion_type"),
+          "You should specify the fusion type");
+    fusion_type_ = static_cast<FusionType>(
+        OperatorBase::GetSingleArgument<int>("fusion_type", FUSION_UNKNOWN));
+    OPERATOR_NEEDS_FEATURE(
+        fusion_type_ > FUSION_UNKNOWN && fusion_type_ < FUSION_MAX,
+        "Undefined Conv fusion type.",
+        fusion_type_);
+
+    if (fusion_type_ == FUSION_CONV_RELU) {
```
Comment (BIT-silence) [inline caffe2/ideep/operators/conv_op.cc]: Consider using switch case here ?

File: caffe2/ideep/operators/conv_pool_base_op.h [hunk 20]
```diff
@@ -35,15 +32,15 @@ class IDEEPConvPoolOpBase : public ConvPoolOpBase<IDEEPContext> {
   ideep::tensor::dims CalcOutputDims(
       const ideep::tensor& input,
       int output_channel) {
-    CAFFE_ENFORCE(input.get_descriptor().get_size() > 0);
+    CAFFE_ENFORCE(input.get_size() > 0);
```
Comment (BIT-silence) [inline caffe2/ideep/operators/conv_pool_base_op.h]:
```suggestion
    CAFFE_ENFORCE_GT(input.get_size(), 0);
```

File: caffe2/opt/optimize_ideep.cc [hunk 21]
```diff
@@ -126,19 +123,18 @@ void resetConvForFusion(repr::NNGraph::NodeRef convNode, int fusion_type) {
...
+        CAFFE_ENFORCE(arg.i() == FUSION_CONV_SUM, "Invalid nest fusion");
+        arg.set_i(FUSION_CONV_SUM_RELU);
         return;
       }
     }
     return;
   }
 
-  CAFFE_ENFORCE(fusion_type < 3, "Invalid fusion type");
+  CAFFE_ENFORCE(fusion_type < FUSION_CONV_SUM_RELU, "Invalid fusion type");
```
Comment (BIT-silence) [inline caffe2/opt/optimize_ideep.cc]:
```suggestion
  CAFFE_ENFORCE_LT(fusion_type, FUSION_CONV_SUM_RELU, "Invalid fusion type");
```

File: caffe2/ideep/operators/operator_fallback_ideep.h [hunk 22]
```diff
@@ -82,20 +82,29 @@ class C10_EXPORT IDEEPFallbackOp final : public IDEEPOperator {
...
+          input_share_[i] = false;
         }
-        input_share_[i] = false;
         auto dtensor = BlobGetMutableTensor(local_input_blobs_[i], CPU);
         dtensor->Resize(input.get_dims());
-        if (input.is_public_format()) {
+        // If fallback from INT8, the public format of original input is nhwc.
+        // While the required format is nchw, need to reorder to nchw.
+        if (input.get_public_format() == iformat::nhwc) {
+          itensor temp_ten ({input.get_dims(), idtype::f32, iformat::nchw},
```
Comment (BIT-silence) [inline caffe2/ideep/operators/operator_fallback_ideep.h]: Please fix the indent here.

File: caffe2/ideep/operators/conv_op.cc [hunk 23]
```diff
@@ -2,116 +2,252 @@
...
+          dummy_scale_,
+          dummy_scale_,
+          attr_,
+          algo_,
+          pk_);
+    }
+
+    if (fusion_type_ == FUSION_CONV_SUM
+        && fusion_type_ == FUSION_CONV_SUM_RELU) {
+      CAFFE_ENFORCE(Y == &(Input(InputSize() - 1)),
```
Comment (BIT-silence) [inline caffe2/ideep/operators/conv_op.cc]:
```suggestion
      CAFFE_ENFORCE_EQ(Y,  &(Input(InputSize() - 1)),
```

File: caffe2/ideep/operators/conv_op.cc [hunk 24]
```diff
@@ -2,116 +2,252 @@
...
+    auto Y_dims_conv = CalcOutputDims(
+        X,
+        grouped ? (filter.get_dim(0) * filter.get_dim(1)) : filter.get_dim(0));
 
     CAFFE_ENFORCE(4 == X.ndims());
-    CAFFE_ENFORCE(4 == filter.ndims());
-    CAFFE_ENFORCE(filter.get_dim(2) == kernel_h());
-    CAFFE_ENFORCE(filter.get_dim(3) == kernel_w());
+    CAFFE_ENFORCE(4 == filter.ndims() || (grouped && (group_ > 1)));
+    CAFFE_ENFORCE(filter.get_dim(2 + grouped) == kernel_h());
```
Comment (BIT-silence) [inline caffe2/ideep/operators/conv_op.cc]:
```suggestion
    CAFFE_ENFORCE_EQ(filter.get_dim(2 + grouped),  kernel_h());
```

File: caffe2/ideep/operators/conv_op.cc [hunk 25]
```diff
@@ -2,116 +2,252 @@
...
+    CAFFE_ENFORCE(filter.get_dim(3 + grouped) == kernel_w());
```
Comment (BIT-silence) [inline caffe2/ideep/operators/conv_op.cc]:
```suggestion
    CAFFE_ENFORCE_EQ(filter.get_dim(3 + grouped), kernel_w());
```

**Comment by gujinghui** [comment]
@yinghai  @BIT-silence 
Corrected the code style.
Pls help review.

**Review by yinghai — APPROVED** [review]
Thanks.

**Pull request closed without merging**